File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/practice/ResolvedLeagueMachineMappings.kt
package com.pillyliu.pinprofandroid.practice

import android.util.Log
import com.pillyliu.pinprofandroid.library.LibraryGameLookup
import org.json.JSONObject

private const val LEAGUE_MACHINE_MAPPINGS_TAG = "PinballDataIntegrity"

internal data class LeagueMachineMappingRecord(
    val machine: String,
    val practiceIdentity: String?,
    val opdbId: String?,
)

internal fun parseLeagueMachineMappings(text: String): Map<String, LeagueMachineMappingRecord> {
    val root = runCatching { JSONObject(text.trim()) }.getOrNull() ?: return emptyMap()
    if (root.optInt("version", 0) < 1) return emptyMap()
    val items = root.optJSONArray("items") ?: return emptyMap()
    val out = linkedMapOf<String, LeagueMachineMappingRecord>()
    for (index in 0 until items.length()) {
        val obj = items.optJSONObject(index) ?: continue
        val machine = obj.optString("machine").trim()
        if (machine.isBlank()) continue
        val key = LibraryGameLookup.normalizeMachineName(machine)
        if (key.isBlank()) continue
        out[key]?.let { existing ->
            Log.w(
                LEAGUE_MACHINE_MAPPINGS_TAG,
                "Duplicate league machine mapping for normalized key $key; replacing ${existing.machine} with $machine",
            )
        }
        out[key] = LeagueMachineMappingRecord(
            machine = machine,
            practiceIdentity = obj.optString("practice_identity").trim().ifBlank { null },
            opdbId = obj.optString("opdb_id").trim().ifBlank { null },
        )
    }
    return out
}

File: Pinball App 2/Pinball App 2/practice/ResolvedLeagueMachineMappings.swift
import Foundation

struct LeagueMachineMappingRecord: Decodable {
    let machine: String
    let practiceIdentity: String?
    let opdbID: String?

    enum CodingKeys: String, CodingKey {
        case machine
        case practiceIdentity = "practice_identity"
        case opdbID = "opdb_id"
    }
}

private struct LeagueMachineMappingsRoot: Decodable {
    let version: Int
    let items: [LeagueMachineMappingRecord]
}

func parseLeagueMachineMappings(text: String) -> [String: LeagueMachineMappingRecord] {
    guard let data = text.data(using: .utf8) ?? text.data(using: .unicode) else { return [:] }
    let decoder = JSONDecoder()
    guard let root = try? decoder.decode(LeagueMachineMappingsRoot.self, from: data), root.version >= 1 else {
        return [:]
    }

    var out: [String: LeagueMachineMappingRecord] = [:]
    for record in root.items {
        let key = LibraryGameLookup.normalizeMachineName(record.machine)
        guard !key.isEmpty else { continue }
        out[key] = record
    }
    return out
}
